content/purgatorio/launch-and-mantain-the-system/monitoring-usage-and-behavior.md

```diff
@@ -1,5 +1,142 @@
+
+With these two resources, you should understand broadly the reasons and challenges of monitoring the Machine Learning models in production, and you should have reasonable strategies to tackle them. 
+
+As you know, the two main challenges regarding the model monitoring are:
+
+- How does the system behave?
+- How is the system used?
+
+Let's see some specific resources for these two challenges.
+
+## Monitoring the Behavior
+
+When monitoring the behavior of the ML model in production, you should consider many aspects:
+
+
+- **Data in:**
+
+    - Monitor whether the data you're processing looks like the data you trained on. ([data drift challenge](https://www.cmswire.com/digital-experience/data-drift-what-it-is-and-how-to-avoid-it/)) E.g., use simple (comparatively interpretable) distributional models to try to track whether data looks "sufficiently" similar.
+    
+    See:
+    - [Calculate Similarity — the most relevant Metrics in a Nutshell](https://towardsdatascience.com/calculate-similarity-the-most-relevant-metrics-in-a-nutshell-9a43564f533e)
+    - [How to measure similarity or dissimilarity between two data set?](https://www.researchgate.net/post/How_to_measure_similarity_or_dissimilarity_between_two_data_set)
+
+
+- **Addressing data drift over time:**
+    
+    - Is your data distribution non-stationary? I.e. are you expecting your model to degrade due to the data changing over time?
+
+    - If so you can do anomaly detection on the stream and track the fractions of anomalous data points over time.
+    
+    See:
+    - [A Gentle Introduction to Concept Drift in Machine Learning](https://machinelearningmastery.com/gentle-introduction-concept-drift-machine-learning/)
+    - [Addressing feature drift in data streams using iterative subset selection](https://dl.acm.org/doi/10.1145/3325061.3325063)
+
+
+- **Data out:**
+
+    - Distribution of predictions by the label. If you see this shifting a lot, another flag that inference data is shifting, which could be a point of concern.
+
+    - Distribution of predictions by confidence/probability. There is a pretty rich (& ongoing) field of research here, but, as a starter, I'd also expect the distribution of raw probabilities (logits) to look similar.
+
+    - As a corollary, looking at overall change in the confidence of outputs. E.g., a drop in confidence values is possibly a concerning sign. (Of course, vanilla NNs are also prone to be highly overconfident on new/OOD data, so you need to be cautious; if this is really critical, could look at different NN variants that generate better confidence assessments.)
+
+    - Consider whether training a model w/ out-of-domain detector (OOD) makes sense or not.
+
+    See: 
+    - [How to determine the confidence of a neural network prediction?](https://stats.stackexchange.com/questions/247551/how-to-determine-the-confidence-of-a-neural-network-prediction)
```

Comment:
To add to the above points, the links and points shared previously also apply. It's also a sign that when we see such changes that our last test/evaluation dataset integrity is failing. Indirectly, it means update the test/evaluation datasets from anomalised production data and retrain the model: either fully or incrementally. 

Or use active-learning to update the model as new data comes into production.

Basically, the steps are plenty to take based on what is the priority and how much you can afford staying outdates from production data?